File: exo2.4.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/wait.h>
#include <errno.h>


/*Ecrire un programme avec un processus pere qui engendre 5 fils dans une bouble for.
Les fils sont nommes fils 1 a fils 5.
Le fils 1 doit afficher "je suis le fils 1" et le fils 2
doit afficher je suis le fils 2, et ainsi de suite.
Indication. on pourra utiliser une variable globale.*/

int numFils = 0;

int main () {

    int i;
    pid_t pid;

    for (i = 1; i <= 5; i++) {
        numFils = i;
        pid = fork ();
        if (pid == -1) {
            fprintf (stderr, "Erreur dans fork () !\n");
            exit (errno);
        }
        if (pid == 0) {
            fprintf (stdout, "Je suis le fils %d.\n", numFils);
            exit (0);
        }
    }

    if (pid == 0) {
        exit (0);
    } else {
        for (i = 1; i <= 5; i++) {
            wait (NULL);
        }
        fprintf (stdout, "Je suis le pere.\n");
    }

    return 0;
}

File: exo2.4.py
import os
import sys

# Programme avec un processus père qui engendre 5 fils
# Les fils sont nommés fils 1 à fils 5
# Utilisation d'une variable globale (simulée par une variable d'instance)

num_fils = 0

def main():
    global num_fils
    pid = None
    
    for i in range(1, 6):
        num_fils = i
        pid = os.fork()
        if pid == -1:
            print("Erreur dans fork() !", file=sys.stderr)
            sys.exit(1)
        if pid == 0:
            print(f"Je suis le fils {num_fils}.")
            sys.exit(0)
    
    if pid == 0:
        sys.exit(0)
    else:
        # Le père attend ses 5 fils
        for _ in range(5):
            os.wait()
        print("Je suis le pere.")

if __name__ == "__main__":
    main()
